### PR TITLE
Esdt safe

### PR DESCRIPTION
**Pull request by dorin-iancu**
ESDT safe for bridge. Tokens that are sent to another chain are locked into this contract. Once the transaction has been executed, the tokens are burned. If, on the other hand, the transaction has been rejected, the tokens are returned to the original sender.

### CODE REVIEW

File: esdt-safe/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    fn remove_token_from_whitelist(&self, token_identifier: TokenIdentifier) -> SCResult<()> {
+        only_owner!(self, "only owner may call this function");
+
+        self.token_whitelist().remove(&token_identifier);
+
+        Ok(())
+    }
+
+    #[endpoint(getNextPendingTransaction)]
+    fn get_next_pending_transaction(&self) -> SCResult<OptionalResult<Transaction<BigUint>>> {
```
Comment (sasurobert) [inline esdt-safe/src/lib.rs:74]: why do we need the pending transactions ?

File: esdt-safe/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        let deposit_remaining = &caller_deposit - &amount;
+        self.send().direct_egld(&caller, &amount, b"whitdrawal");
+        self.set_deposit(&caller, &deposit_remaining);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint(createTransaction)]
+    fn create_transaction(
```
Comment (sasurobert) [inline esdt-safe/src/lib.rs:171]: what does this do ? the Safe contract only needs to lock, burn and send back tokens if the bridge failed. BURN and SEND BACK can be called by whitelisted address | smart contract owner only.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: esdt-safe/src/lib.rs (hunk 2 above)
Comment (dorin-iancu) [reply esdt-safe/src/lib.rs:171]: This endpoint is called by accounts that want to transfer to another chain. A transaction struct is created and saved with the required data (sender, destination, ESDT token used and amount paid), which is saved as a "pending" transaction. The tokens are locked in the contract after this.

File: esdt-safe/src/lib.rs (hunk 1 above)
Comment (dorin-iancu) [reply esdt-safe/src/lib.rs:74]: These are the transactions created by accounts that want to transfer to another chain. When they create a transaction, their tokens are locked inside the contract. Then, the owner can query the contract and get the pending transactions, execute them on the other chain, and then set their status to Executed (which burns the locked tokens) or Rejected (which will refund the tokens to the original sender).

**Pull request merged**